File: practice/leetcode/python/145_binary_tree_postorder_traversal.py
#https://leetcode.com/problems/binary-tree-postorder-traversal/description/
#Given the root of a binary tree, return the postorder traversal of its nodes' values.

from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def postOrderTraversal_Iterative(self, root: Optional[TreeNode]) -> List[int]:
        answer = []
        stack = []
        last = None
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                temp = stack.pop()
                if temp.right != None and temp.right != last:
                    root = temp.right
                else:
                    answer.append(temp.val)
                    last = temp
                    stack.pop()
        return(answer)

File: practice/leetcode/cpp/145_binary_tree_postorder_traversal.cpp
//https://leetcode.com/problems/binary-tree-postorder-traversal/description/
//Given the root of a binary tree, return the postorder traversal of its nodes' values.

#include <queue>
#include <stack>
#include <vector>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(): val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x): val(x), left(nullptr), right(nullptr) {}
};

class Solution {
private:
	void postOrder(TreeNode *root, vector<int>	&answer) {
		if (!root) {
			return;
		}
		postOrder(root->left, answer);
		postOrder(root->right, answer);
		answer.push_back(root->val);
		return;
	}
public:
	vector<int> postOrder_Recurse(TreeNode *root) {
		vector<int> answer;
		if (!root) {
			return answer;
		}
		postOrder(root, answer);
		return(answer);
	}
	vector<int> postOrder_Iterative(TreeNode *root) {
		stack<TreeNode *> st;
		vector<int> answer;
		TreeNode *last = nullptr, *temp = nullptr;

		if (!root) {
			return answer;
		}
		//iterative loop
		while (root || !st.empty()) {
			if (root) {
				st.push(root);
				root = root->left;
			} else {
				temp = st.top();
				//
				if ((temp->right != nullptr) && (temp->right != last)) {
					root = temp->right;
				} else {
					answer.push_back(temp->val);
					last = temp;
					st.pop();
				}
			}
		}
		return(answer);
	}
};
